README.md
# Coast to Coast

Coast to Coast is a remake of the Apple IIe game "How the West Was One + Three X Four" which taught grade school students how to remember mathematical order of operations.

A live demo is available at http://coast-to-coast.snyderc.me/

## Installing

Coast to Coast is a JavaScript game and will run in a modern web browser that supports async/await. Open game.html to begin playing. The game uses several JavaScript files, listed in index.html, and uses the images in the "images" folder.

## Gameplay

You are racing your computer opponent to see whose car can travel across the country first. (Which country? Pick your favorite!) The country is represented by a grid from 0-50, where position 0 is the start and position 50 is the end.

On each turn, you are presented with three random integers -- one in the range 1-3, one in the range 0-4, and one in the range 1-7 (all ranges are inclusive). Using all three numbers, and the provided math symbols (one each of addition, subtraction, multiplication, division, and a set of parentheses), your goal is to come up with the equation whose answer gives you the best possible position on the gameboard.

The goal is to end your turn on position 50 without going over. If you would go over on a turn, you do not move that turn. If the player who started a game lands on position 50, the other player has one last turn to attempt to land on position 50 and tie the game.

Your equation can result in a positive or negative number, or in zero. An equation that equals zero is unlikely to give you an advantage, but a negative number can sometimes trigger a special feature and give you an advantage.

There are a few special features:

### Shortcuts

If you end your turn on a shortcut, you get to move ahead to the shortcut's end position.

### Rest Stops

There are four rest stops -- at positions 10, 20, 30, and 40. If you end your turn on a rest stop, you get to move ahead to the next rest stop.

### Landing on Other Player

If you end your turn on the other player's position, the other player gets sent back two rest stops. (If the other player was already at a rest stop, they are safe and do not get sent back.)

## Limitations

This version of Coast to Coast recreates the basic game logic. However, the computer player will always choose the most advantageous move to it, which results in short games if the computer gets lucky with the numbers it is assigned. The games used to go much slower when I was in elementary school and playing against the computer's "average" mode!

The original game includes plentiful age-appropriate explanatory text to walk a child through the gameplay. This version of Coast to Coast has limited explanatory text and has no in-game tutorial.

The original game also had features that allowed a user to get hints, such as whether the user can make a certain value given their three numbers, and what the computer's suggested move would be for a turn. This version of Coast to Coast provides suggested moves by outputting to the JavaScript console, but does not otherwise provide visible in-game hints.

Finally, since this version of Coast to Coast is a demonstration, there are a few bugs that have been identified that will be fixed over time. You can see those bugs in the "bugs.txt" file.

## Author

Chris Snyder

## License

This project is licensed under the MIT License - see the [LICENSE.md](LICENSE.md) file for details.

## Acknowledgements

"How the West Was One + Three X Four" was created by Bonnie Seiler and Pat Sine, (C)1987. No copyrighted elements of the game are used in Coast to Coast.

CHANGELOG.md
# Changelog

## [0.1.0] - 2018-09-11
- Initial release.

messages.js
const messages = {
    gameLoading: {
        type: "text",
        x: 450,
        y: 300,
        value: "Loading..."
    },
    coastToCoastName: {
        type: "text",
        x: 450,
        y: 100,
        value: "Coast to Coast"
    },
    startGameButton: {
        type: "rectangle",
        x: 300,
        y: 150,
        width: 300,
        height: 80,
        value: "Click to Start Game",
        onClick: function() {
            canvas.removeEventListener("click", clickListenerCallback);
            gameVars.textAndButtonsToDraw.length = 0;
            window.requestAnimationFrame(drawScreen);
            // requestUserName(); // Not asking user for their name in version 1 of game
            return initializeGame();
        }
    },
    startGameRedCarDecoration: {
        type: "image",
        x: 220,
        y: 80,
        url: "/images/car-right-red.png"
    },
    startGameBlueCarDecoration: {
        type: "image",
        x: 600,
        y: 80,
        url: "/images/car-left-blue.png"
    },
    rules1: {
        type: "text",
        x: 450,
        y: 350,
        value: "How to Play:"
    },
    rules2: {
        type: "text",
        x: 450,
        y: 400,
        value: "Create an equation that moves your car along the board!"
    },
    rules3: {
        type: "text",
        x: 450,
        y: 440,
        value: "Land on a rest stop and go ahead another rest stop!"
    },
    rules4: {
        type: "text",
        x: 450,
        y: 480,
        value: "Land on another player not at a rest stop"
    },
    rules5: {
        type: "text",
        x: 450,
        y: 520,
        value: "and bump them back two rest stops!"
    },
    rules6: {
        type: "text",
        x: 450,
        y: 560,
        value: "Land right on the rest stop at position 50 to win!"
    },
    whatIsYourNameText: {
        type: "text",
        x: 450,
        y: 200,
        value: "Type your name then press Enter"
    },
    whatIsYourNameInputBox: {
        type: "rectangle",
        x: 300,
        y: 300,
        width: 300,
        height: 80,
    },
    whatIsYourNameUserInput: {
        type: "text",
        x: 450,
        y: 350,
        value: ""
    },
    welcomeMessageRed: {
        type: "text",
        x: 450,
        y: 40,
        value: `Welcome! You will have the red car.`
    },
    welcomeMessageBlue: {
        type: "text",
        x: 450,
        y: 80,
        value: "The computer will have the blue car."
    },
    newGameMessageRed: {
        type: "text",
        x: 450,
        y: 120,
        value: "You will go first."
    },
    newGameMessageBlue: {
        type: "text",
        x: 450,
        y: 80,
        value: "The computer will go first."
    },
    nextTurnMessageHuman: {
        type: "text",
        x: 450,
        y: 80,
        value: "It is your turn."
    },
    nextTurnMessageComputer: {
        type: "text",
        x: 450,
        y: 80,
        value: "It is the computer's turn."
    },
    startTurnButton: {
        type: "rectangle",
        x: 300,
        y: 140,
        width: 300,
        height: 50,
        name: "Start Turn button",
        value: "Click to Continue",
        onClick: function() {
            canvas.removeEventListener("click", clickListenerCallback);
            gameVars.textAndButtonsToDraw.length = 0;
            window.requestAnimationFrame(drawScreen);
            return startTurn();
        }
    },
    turnButtons: [
        {
            type: "rectangle",
            x: 110,
            y: 20,
            width: 340,
            height: 50,
            name: "Equation",
            value: ""
        },
        {
            type: "rectangle",
            x: 110,
            y: 80,
            width: 680,
            height: 50,
            name: "Create Your Equation text",
            value: "Create Your Equation:"
        },
        {
            type: "rectangle",
            x: 110,
            y: 140,
            width: 50,
            height: 50,
            name: "First Number Button",
            value: -1,
            onClick: function(i) { handleEquationButtonClick(this, i) },
            clicked: false
        },
        {
            type: "rectangle",
            x: 170,
            y: 140,
            width: 50,
            height: 50,
            name: "Second Number Button",
            value: -1,
            onClick: function(i) { handleEquationButtonClick(this, i) },
            clicked: false
        },
        {
            type: "rectangle",
            x: 230,
            y: 140,
            width: 50,
            height: 50,
            name: "Third Number Button",
            value: -1,
            onClick: function(i) { handleEquationButtonClick(this, i) },
            clicked: false
        },
        {
            type: "rectangle",
            x: 290,
            y: 140,
            width: 50,
            height: 50,
            value: "+",
            onClick: function(i) { handleEquationButtonClick(this, i) },
            clicked: false
        },
        {
            type: "rectangle",
            x: 350,
            y: 140,
            width: 50,
            height: 50,
            value: "-",
            onClick: function(i) { handleEquationButtonClick(this, i) },
            clicked: false
        },
        {
            type: "rectangle",
            x: 410,
            y: 140,
            width: 50,
            height: 50,
            value: "*",
            onClick: function(i) { handleEquationButtonClick(this, i) },
            clicked: false
        },
        {
            type: "rectangle",
            x: 470,
            y: 140,
            width: 50,
            height: 50,
            value: "/",
            onClick: function(i) { handleEquationButtonClick(this, i) },
            clicked: false
        },
        {
            type: "rectangle",
            x: 530,
            y: 140,
            width: 50,
            height: 50,
            value: "(",
            onClick: function(i) { handleEquationButtonClick(this, i) },
            clicked: false
        },
        {
            type: "rectangle",
            x: 590,
            y: 140,
            width: 50,
            height: 50,
            value: ")",
            onClick: function(i) { handleEquationButtonClick(this, i) },
            clicked: false
        },
        {
            type: "rectangle",
            x: 650,
            y: 140,
            width: 50,
            height: 50,
            value: "=",
            onClick: function(i) { handleEquationButtonClick(this, i) },
            clicked: false
        },
        {
            type: "rectangle",
            x: 710,
            y: 140,
            width: 80,
            height: 50,
            value: "Clear",
            onClick: function(i) { handleEquationButtonClick(this, i) },
            clicked: false
        }
    ],
    equalsSignDisplay: {
        type: "rectangle",
        x: 450,
        y: 20,
        width: 50,
        height: 50,
        value: "="
    },
    answerInput: {
        type: "rectangle",
        x: 500,
        y: 20,
        width: 290,
        height: 50,
        value: ""
    },
    answerPrompt: {
        type: "rectangle",
        x: 110,
        y: 80,
        width: 680,
        height: 50,
        value: "Type your answer above and press Enter"
    },
    incorrectAnswer: {
        type: "rectangle",
        x: 110,
        y: 80,
        width: 680,
        height: 50,
        value: "Incorrect. Press Enter."
    },
    displayCorrectEquation: {
        type: "rectangle",
        x: 110,
        y: 140,
        width: 680,
        height: 50,
        value: ""
    },
    shortcutModifierDisplay: {
        type: "text",
        x: 450,
        y: 40,
        value: "Shortcut! Move ahead!"
    },
    landmarkModifierDisplay: {
        type: "text",
        x: 450,
        y: 40,
        value: "Reached a rest stop! Move ahead to next rest stop!"
    },
    bumpedBackModifierDisplay: {
        type: "text",
        x: 450,
        y: 80,
        value: "Landed on same spot as other player!"
    },
    successfulEndOfTurnMessage: {
        type: "text",
        x: 450,
        y: 120,
        value: "Press Enter to continue."
    },
    computerTurnRandomNumbersDisplay: {
        type: "text",
        x: 450,
        y: 40,
        value: ""
    },
    endGameTieMessage: {
        type: "text",
        x: 450,
        y: 40,
        value: "Both players tie!"
    },
    endGameHumanWins: {
        type: "text",
        x: 450,
        y: 40,
        value: "You win!"
    },
    endGameComputerWins: {
        type: "text",
        x: 450,
        y: 40,
        value: "The computer wins!"
    },
    endGameNewGameButton: {
        type: "rectangle",
        x: 350,
        y: 140,
        width: 200,
        height: 50,
        value: "New Game",
        onClick: function() {
            canvas.removeEventListener("click", clickListenerCallback);
            gameVars.textAndButtonsToDraw.length = 0;
            window.requestAnimationFrame(drawScreen);
            return initializeGame();
        }
    },
}

gameEngine.js
// Game Engine:
// On load, call loadAssets()
// When loadAssets done, it resolves promise, triggering showStartGameButton
// When user completes showStartGameButton (clicks on button), go to requestUserName
// When user completes requestUserName (clicks on button), go to initializeGame
// When initializeGame done, it calls showStartTurnButton
// When user completes initializeTurn (clicks on button), go to startTurn
// startTurn either calls startHumanTurn or startComputerTurn
// showTurnButtons -> showAnswerPrompt -> processMove. processComputerTurn -> processMove
// When user completes showTurnButtons (clicks equal sign), go to showAnswerPrompt
// When user completes showAnswerPrompt (presses Enter), go to processMove
// When processMove done, it either calls:
// -- showStartTurnButton if win condition has not been met
// -- endGame if win condition has been met
// User can complete endGame by clicking "New Game" -- goes to initializeGame()

// Game Graphics:
// During game, "Board" (gameboard) and "Display Area" (text, buttons, etc.)

// Game state is kept here
const gameVars = {
    pregame: true,             // Screen will only have board painted if pregame === false
    textAndButtonsToDraw: [],  // Array of elements to draw in Display Area (non-board area) or during pregame
    landmarks: [
      10,
      20,
      30,
      40
    ],
    finish: 50,
    shortcuts: [    // Hard-coded now. Display is also hard-coded.
      {
        start: 5,
        end: 14
      },
      {
        start: 26,
        end: 35
      }
    ],
    // Turn Variables
    turnRandomNumbers: [],
    turnEntryStack: [],
    currentPlayer: undefined,
    otherPlayer: undefined,
    // Game Variables
    numberOfGamesPlayed: 0,
    goesFirst: undefined,
    gameOver: false,
    // Players
    playerOne: {
        name: undefined,
        score: 0,
        type: 'human',
        carName: 'redCar'
    },
    playerTwo: {
        name: 'Blue Racer',
        score: 0,
        type: 'computer',
        carName: 'blueCar'
    },
    // Cars
    carPositions: {
        redCar: {
            x: -50,
            y: -50,
            futurePositions: [],
            dir: 'right'
        },
        blueCar: {
            x: -50,
            y: -50,
            futurePositions: [],
            dir: 'right'
        }
    }
}

// Global constants

const canvas = document.getElementById("myCanvas");
const ctx = canvas.getContext("2d");
const images = {};
const dictOfCarCoords = {};

// Getters and Setters for gameVars

function setPlayerOneName(name) {
    gameVars.playerOne.name = name;
}
  
function setTurnRandomNumbers() {
    // First, clear out existing random numbers
    gameVars.turnRandomNumbers.length = 0;
    // first random number: 1-3
    gameVars.turnRandomNumbers.push(Math.floor(Math.random()*3+1));
    // second random number: 0-4
    gameVars.turnRandomNumbers.push(Math.floor(Math.random()*5));
    // third random number: 1-7
    gameVars.turnRandomNumbers.push(Math.floor(Math.random()*7+1));
}

function clearTurnRandomNumbers() {
    gameVars.turnRandomNumbers.length = 0;
}

function clearTurnEntryStack() {
    gameVars.turnEntryStack.length = 0;
}

//
// Game Engine functions
//

function loadAssets() {
    // Draw "Loading..." text on canvas
    gameVars.textAndButtonsToDraw.push(messages.gameLoading);
    window.requestAnimationFrame(drawScreen);

    // Initializes car coordinate dictionary
    initializeCarCoords();

    // Loads images from files, then callback is showStartGameButton
    loadImages(imageUrls, images, function() {
        // Clear text/button array and call showStartGameButton
        gameVars.textAndButtonsToDraw.length = 0;
        showStartGameButton();
    });
}

function showStartGameButton() {
    // Draw Coast to Coast name and start game button
    gameVars.textAndButtonsToDraw.push(messages.coastToCoastName);
    gameVars.textAndButtonsToDraw.push(messages.startGameButton);
    gameVars.textAndButtonsToDraw.push(messages.startGameRedCarDecoration);
    gameVars.textAndButtonsToDraw.push(messages.startGameBlueCarDecoration);
    gameVars.textAndButtonsToDraw.push(messages.rules1);
    gameVars.textAndButtonsToDraw.push(messages.rules2);
    gameVars.textAndButtonsToDraw.push(messages.rules3);
    gameVars.textAndButtonsToDraw.push(messages.rules4);
    gameVars.textAndButtonsToDraw.push(messages.rules5);
    gameVars.textAndButtonsToDraw.push(messages.rules6);
    // Add click event listener to canvas
    canvas.addEventListener('click', clickListenerCallback);
    // Draw text/buttons
    window.requestAnimationFrame(drawScreen);
}

// Version 1 of game won't customize the display with the user's name.
// function requestUserName() {
//     // Queue elements to draw to screen
//     gameVars.textAndButtonsToDraw.push(messages.whatIsYourNameText);
//     gameVars.textAndButtonsToDraw.push(messages.whatIsYourNameInputBox);
//     messages.whatIsYourNameUserInput.value = "";
//     gameVars.textAndButtonsToDraw.push(messages.whatIsYourNameUserInput);
//     // Add keypress listener to window
//     window.addEventListener('keydown', function _keydownListener(event) {
//         keydownListenerCallback(event, messages.whatIsYourNameUserInput, "letters", 15, function(name) {
//             setPlayerOneName(name.trim());
//             gameVars.textAndButtonsToDraw.length = 0;
//             window.removeEventListener('keydown', _keydownListener);
//             initializeGame();
//         });
//     });
//     // Draw text/buttons
//     // Once pregame === false, drawing will be handled by a different function
//     if (gameVars.pregame === true) {
//         window.requestAnimationFrame(drawScreen);
//     }
// }

function initializeGame() {
    // If this is the first game, display a welcome message to the user
    if (gameVars.numberOfGamesPlayed === 0) {
        gameVars.pregame = false;
        gameVars.textAndButtonsToDraw.push(messages.welcomeMessageRed);
        gameVars.textAndButtonsToDraw.push(messages.welcomeMessageBlue);
    }

    // Set who goes first (P1 for odd-numbered games; P2 for even-numbered games)
    gameVars.numberOfGamesPlayed++;
    if (gameVars.numberOfGamesPlayed % 2 === 0) {
        gameVars.goesFirst = gameVars.playerTwo;
        gameVars.textAndButtonsToDraw.push(messages.newGameMessageBlue);
    }
    else {
        gameVars.goesFirst = gameVars.playerOne;
        gameVars.textAndButtonsToDraw.push(messages.newGameMessageRed);
    }
    // Set both player scores to 0
    gameVars.playerOne.score = 0;
    gameVars.playerTwo.score = 0;

    // Reset car positions
    gameVars.carPositions.redCar.x = dictOfCarCoords[gameVars.playerOne.score].redCar.x;
    gameVars.carPositions.redCar.y = dictOfCarCoords[gameVars.playerOne.score].redCar.y;
    gameVars.carPositions.blueCar.x = dictOfCarCoords[gameVars.playerTwo.score].blueCar.x;
    gameVars.carPositions.blueCar.y = dictOfCarCoords[gameVars.playerTwo.score].blueCar.y;

    // Reset currentPlayer and otherPlayer
    gameVars.currentPlayer = undefined;
    gameVars.otherPlayer = undefined;

    // TODO (future feature): Clear shortcut array and set array of shortcut start and end points from a list of constraints
    // NOTE: When you implement that feature, also change the drawShortcuts function so it draws the points dynamically

    // Paint screen (which will now show board b/c no longer in pregame)
    window.requestAnimationFrame(drawScreen);
    return initializeTurn();
}

function initializeTurn() {
    // If beginning of game, set the current player to goesFirst
    if (gameVars.currentPlayer === undefined) {
        gameVars.currentPlayer = gameVars.goesFirst;
        if (gameVars.playerOne !== gameVars.currentPlayer) {
            gameVars.otherPlayer = gameVars.playerOne;
        }
        else {
            gameVars.otherPlayer = gameVars.playerTwo;
        }
    }
    else {
        // Before you swap the current player and other player, check win condition first
        // Win Condition:
        // If player who goes first reaches 50, other player has 1 turn to also reach 50.
        // If the current player from previous turn reached 50, and they didn't go first, they win
        // If the other player has score 50, and current player didn't reach a score of 50, other player wins
        // Otherwise, gameplay continues
        if (gameVars.currentPlayer.score === 50 && gameVars.otherPlayer.score === 50) {
            return endGame("tie");
        }
        // If current player didn't go first, current player wins
        else if (gameVars.currentPlayer.score === 50 && gameVars.currentPlayer != gameVars.goesFirst) {
            return endGame("win", gameVars.currentPlayer);
        }
        else if (gameVars.otherPlayer.score === 50) {
            // If other player already had 50 points, and current player didn't get to 50 last turn, other player wins
            return endGame("win", gameVars.otherPlayer);
        }

        // Swap current player and other player
        const tempPlayer = gameVars.currentPlayer;
        gameVars.currentPlayer = gameVars.otherPlayer;
        gameVars.otherPlayer = tempPlayer;

        // Show a message about whose turn it is
        if (gameVars.currentPlayer.type === "human") {
            gameVars.textAndButtonsToDraw.push(messages.nextTurnMessageHuman);
        }
        else if (gameVars.currentPlayer.type === "computer") {
            gameVars.textAndButtonsToDraw.push(messages.nextTurnMessageComputer);
        }
    }

    // Reset turn variables
    clearTurnEntryStack();
    setTurnRandomNumbers();
    gameVars.correctEquationAndAnswer = undefined;

    gameVars.textAndButtonsToDraw.push(messages.startTurnButton);

    canvas.addEventListener("click", clickListenerCallback);

    window.requestAnimationFrame(drawScreen);
}

function startTurn() {
    function startComputerTurn() {
        // console.log("In startComputerTurn()");
        // Display the 3 random numbers the computer is working with
        messages.computerTurnRandomNumbersDisplay.value = `Computer's numbers: ${gameVars.turnRandomNumbers}`;
        gameVars.textAndButtonsToDraw.push(messages.computerTurnRandomNumbersDisplay);
        window.requestAnimationFrame(drawScreen);
        const result = runComputerAi(gameVars.turnRandomNumbers, gameVars.currentPlayer.score, gameVars.otherPlayer.score);
        // console.log(gameVars.turnRandomNumbers);
        // console.log(result);
        messages.displayCorrectEquation.value = `${result.formulaForMove.formula} = ${result.formulaForMove.result}`;
        processMove(result.formulaForMove.formula, result.formulaForMove.result);
    }

    function startHumanTurn() {
        // console.log("In startHumanTurn()");
        function showTurnButtons() {
            for (let i = 0; i < messages.turnButtons.length; i++) {
                // Set the three numbers to be used on this turn, from the array turnRandomNumbers
                if (messages.turnButtons[i].name === "First Number Button") {
                    messages.turnButtons[i].value = gameVars.turnRandomNumbers[0].toString();
                }
                else if (messages.turnButtons[i].name === "Second Number Button") {
                    messages.turnButtons[i].value = gameVars.turnRandomNumbers[1].toString();
                }
                else if (messages.turnButtons[i].name === "Third Number Button") {
                    messages.turnButtons[i].value = gameVars.turnRandomNumbers[2].toString();
                }
                gameVars.textAndButtonsToDraw.push(messages.turnButtons[i]);
            }
            addEventListener("click", clickListenerCallback);
            window.requestAnimationFrame(drawScreen);
        }
        showTurnButtons();
    }

    // Call either startComputerTurn or startHumanTurn depending on the player
    if (gameVars.currentPlayer.type === "human") {
        return startHumanTurn();
    }
    else if (gameVars.currentPlayer.type === "computer") {
        return startComputerTurn();
    }
}

async function processMove(equation, answer) {
    moveOutcome = getMoveOutcome(gameVars.currentPlayer.score, gameVars.otherPlayer.score, answer);
    // console.log(moveOutcome);
    gameVars.textAndButtonsToDraw.push(messages.displayCorrectEquation); // Displays the equation
    // Move current player car based on math
    await moveCar(
        gameVars.currentPlayer.carName,
        moveOutcome.curPlayerOriginalPos,
        moveOutcome.curPlayerPositionAfterMath
    )
    // console.log("First move done");
    // Clear out any old values from the computerTurn function, and redisplay just the equation
    gameVars.textAndButtonsToDraw.length = 0;
    gameVars.textAndButtonsToDraw.push(messages.displayCorrectEquation);
    // If modifier, move current player car
    if (moveOutcome.curPlayerTypeOfModifier !== "none") {
        if (moveOutcome.curPlayerTypeOfModifier === "landmark") {
            gameVars.textAndButtonsToDraw.push(messages.landmarkModifierDisplay);
        }
        else if (moveOutcome.curPlayerTypeOfModifier === "shortcut") {
            gameVars.textAndButtonsToDraw.push(messages.shortcutModifierDisplay);
        }
        await moveCar(
            gameVars.currentPlayer.carName,
            moveOutcome.curPlayerPositionAfterMath,
            moveOutcome.curPlayerFinalPosAfterModifier,
            moveOutcome.curPlayerTypeOfModifier
        )
        // console.log("Second move done");
    }
    // If other player gets displaced, move other player car
    if (moveOutcome.isOtherPlayerDisplaced) {
        gameVars.textAndButtonsToDraw.push(messages.bumpedBackModifierDisplay);
        await moveCar(
            gameVars.otherPlayer.carName,
            moveOutcome.otherPlayerOriginalPos,
            moveOutcome.otherPlayerFinalPos
        )
        // console.log("Third move done");
    }
    // console.log("All moves done");
    // Update scores
    gameVars.currentPlayer.score = moveOutcome.curPlayerFinalPosAfterModifier;
    gameVars.otherPlayer.score = moveOutcome.otherPlayerFinalPos;
    // Prompt user to press Enter to continue. Pressing Enter clears the text display area's array.
    gameVars.textAndButtonsToDraw.push(messages.successfulEndOfTurnMessage);
    window.requestAnimationFrame(drawScreen);
    window.addEventListener("keydown", endTurn);
}

function endGame(outcome, winner = undefined) {
    // console.log("In endGame");
    if (outcome == "tie") {
        gameVars.textAndButtonsToDraw.push(messages.endGameTieMessage);
    }
    else if (winner.type === "human") {
        gameVars.textAndButtonsToDraw.push(messages.endGameHumanWins);
    }
    else if (winner.type === "computer") {
        gameVars.textAndButtonsToDraw.push(messages.endGameComputerWins);
    }
    gameVars.textAndButtonsToDraw.push(messages.endGameNewGameButton);
    canvas.addEventListener("click", clickListenerCallback);
}

loadAssets();